updater/sponsorships.py

```diff
@@ -50,82 +50,131 @@ def get_events(app):
 
 
+def applyV5(app, context_id):
+    # check the context id is linked
+    context = contexts[app['context']]
+    collection = context['collection']
+    c = db[collection].find({'contextId': context_id})
+    if c.empty():
+        db['sponsorships'].insert({
+            '_from': 'users/0',
+            '_to': 'apps/' + app['_key'],
+            'expireDate': int(time.time()) + 3600,
+            'contextId': context_id,
+        })
+        print(f"{context_id} is not linked to a brightid yet, a temporary sponsorship is added to be applied after user links contextId")
+        return
+
+    user = c.next()['user']
+    c = sponsorships.find({'_from': f'users/{user}'})
+    if not c.empty():
+        print("the user is sponsored before")
+        return
+
+    op = {
+        'name': 'Sponsor',
+        'app': app['_key'],
+        'id': user,
+        'timestamp': int(time.time() * 1000),
+        'v': 5
+    }
+
+    signing_key = ed25519.SigningKey(
+        base64.b64decode(app['sponsorPrivateKey']))
+    message = json.dumps(op, sort_keys=True,
+                         separators=(',', ':')).encode('ascii')
+    sig = signing_key.sign(message)
+    op['sig'] = base64.b64encode(sig).decode('ascii')
+    h = base64.b64encode(sha256(message).digest()).decode("ascii")
+    op['hash'] = h.replace('/', '_').replace('+', '-').replace('=', '')
+    op['state'] = 'init'
+    op['_key'] = op['hash']
+    operation = db['operations'].get(op['hash'])
+    if not operation:
+        db['operations'].insert(op)
+
+
+def applyV6(app, app_id):
+    c = sponsorships.find({'appId': app_id})
+    if c.empty():
+        db['sponsorships'].insert({
+            '_from': 'users/0',
+            '_to': 'apps/' + app['_key'],
+            'expireDate': int(time.time()) + 3600,
+            'appId': app_id,
+            'state': 'app',
+        })
+        return
+
+    sponsorship = c.next()
+    if sponsorship['state'] in ['app', 'done']:
+        print('the app id is sponsored before')
+        return
+
+    db['sponsorships'].update({
+        '_key': sponsorship['_key'],
+        'expireDate': None,
+        'state': 'done',
+        'timestamp': int(time.time() * 1000),
+    })
+
+
+def has_sponsorship(app):
+    tsponsorships = app['totalSponsorships']
+    usponsorships = sponsorships.find(
+                {'_to': 'apps/{0}'.format(app['_key'])}).count()
+    return tsponsorships - usponsorships > 0
+
+
+def remove_testblocks(app, context_id):
+    # remove testblocks if exists
+    tblocks = testblocks.find({
+        'contextId': context_id,
+        'action': 'sponsorship',
+        'app': app['_key']
+    }).batch()
+    for tblock in tblocks:
+        testblocks.delete(tblock)
+
+
+def is_using_sponsor_contract(app):
+    if not app.get('sponsorEventContract'):
+        return False
+    if not app.get('wsProvider'):
+        return False
+    if not app.get('usingBlindSig') and not app.get('sponsorPrivateKey'):
+        return False
```

Comment:
Why should a v5 app need a `sponsorPrivateKey` to create a `Sponsor` event from a smart contract? Also, the `app.sponsorPrivateKey` is never used anywhere, so why do we store it in our app registry?

```suggestion
```

Reply:
In V5 all nodes don't have links of all the apps and cannot apply sponsorships. Apps rely on a specific node that has links and also spending private key to sign sponsor operation that has brightid.

Reply:
What happens if two nodes that have links for an app both receive the sponsor operation? How do they decide which one applies it? Also, why should an app register its sponsor private key? It seems only the sponsor public key is ever used by nodes.

Reply:
> What happens if two nodes that have links for an app both receive the sponsor operation? How do they decide which one applies it?
Apps should select one node as their trusted node but there is no problem if they select more than one, the first operation that has been sent will apply and the other ones will fail.

> It seems only the sponsor public key is ever used by nodes.
In v5 the sponsor private keys are used by nodes in both sponsoring by operation and sponsoring by contracts:
https://github.com/BrightID/BrightID-Node/blob/e57447fca6446fae25f1a107f5f5e35693ee11cd/updater/sponsorships.py#L113
https://github.com/BrightID/BrightID-Node/blob/a1016a97fba96c84949c4675dd2442c2481f70ae/web_services/foxx/brightid/db.js#L841

> Also, can't the app use something else to show the node is allowed to link?
We don't use this private key as a flag if I got what you mean correctly.

> If an app only uses events to sponsor it shouldn't have to create a sponsor key pair.
Events contain only context id and we need the links to find out which users are sponsored in v5.